xds/src/main/java/io/grpc/xds/XdsClientImpl.java

```diff
@@ -426,6 +496,111 @@ private String processRouteConfig(RouteConfiguration config) {
+      String clusterName = assignment.getClusterName();
+      // Skip information for clusters not requested.
+      // Management server is required to always send newly requested resources, even if they
+      // may have been sent previously (proactively). Thus, client does not need to cache
+      // unrequested resources.
+      if (!endpointWatchers.containsKey(clusterName)) {
+        continue;
+      }
+      EndpointUpdate.Builder updateBuilder = EndpointUpdate.newBuilder();
+      updateBuilder.setClusterName(clusterName);
+      if (assignment.getEndpointsCount() == 0) {
+        errorMessage = "Cluster without any locality endpoint.";
+        break;
+      }
+      // The policy.disable_overprovisioning field must be set to true.
+      if (!assignment.getPolicy().getDisableOverprovisioning()) {
+        errorMessage = "Cluster requires overprovisioning.";
+        break;
+      }
+      for (io.envoyproxy.envoy.api.v2.endpoint.LocalityLbEndpoints localityLbEndpoints
+          : assignment.getEndpointsList()) {
+        // The lb_endpoints field for LbEndpoint must contain at least one entry.
+        if (localityLbEndpoints.getLbEndpointsCount() == 0) {
+          errorMessage = "Locality with no endpoint.";
+          break validateData;
+        }
+        // The endpoint field of each lb_endpoints must be set.
+        // Inside of it: the address field must be set.
+        for (io.envoyproxy.envoy.api.v2.endpoint.LbEndpoint lbEndpoint
+            : localityLbEndpoints.getLbEndpointsList()) {
+          if (!lbEndpoint.hasEndpoint() || !lbEndpoint.getEndpoint().hasAddress()) {
+            errorMessage = "Invalid endpoint address information.";
+            break validateData;
+          }
+        }
+        // Note endpoints with health status other than UNHEALTHY and UNKNOWN are still
+        // handed over to watching parties. It is watching parties' responsibility to
+        // filter out unhealthy endpoints. See EnvoyProtoData.LbEndpoint#isHealthy().
+        updateBuilder.addLocalityLbEndpoints(
+            Locality.fromEnvoyProtoLocality(localityLbEndpoints.getLocality()),
+            LocalityLbEndpoints.fromEnvoyProtoLocalityLbEndpoints(localityLbEndpoints));
+      }
+      for (io.envoyproxy.envoy.api.v2.ClusterLoadAssignment.Policy.DropOverload dropOverload
+          : assignment.getPolicy().getDropOverloadsList()) {
+        updateBuilder.addDropPolicy(DropOverload.fromEnvoyProtoDropOverload(dropOverload));
+      }
+      EndpointUpdate update = updateBuilder.build();
+      endpointUpdates.put(clusterName, update);
+    }
+    if (errorMessage != null) {
+      adsStream.sendNackRequest(ADS_TYPE_URL_EDS, adsStream.edsResourceNames,
+          edsResponse.getNonce(),
+          "ClusterLoadAssignment message contains invalid information for gRPC's usage: "
+              + errorMessage);
+      return;
+    }
+    adsStream.sendAckRequest(ADS_TYPE_URL_EDS, adsStream.edsResourceNames,
```

Comment:
In all the other cases you are using `endpointWatchers.keySet()` instead of `adsStream.edsResourceNames`. The `adsStream.sendAckRequest` will just set the same value back to `adsStream.edsResourceNames` so this usage doesn't make sense.

Reply:
Yes, this is something I considered. `adsStream.edsResourceNames` contains the most recently requested resource names, which is always same as cluster names that existing watchers are interested in. Our implementation maintains this invariant. But there is a special case: as we are discussing in https://github.com/grpc/grpc-java/pull/6370#discussion_r348859192, if server violates the protocol by sending an EDS response while the client never initiated, we are not able to catch this breakage without `adsStream.edsResourceNames != null` check. However, as I said, we may want to just handle the protocol violation by ignoring it (it's server's responsibility to do the right thing). So I can change to eliminate `adsStream.edsResourceNames`.